### PR TITLE
Add "Create User" button to admin dashboard

### LINKED ISSUE #666: It should be more obvious to admins how to create users
We initially tried to avoid letting admin create users.
We've realized/decided now that there is a good use case for that.
- Users aren't tech savvy
- Users don't speak English (the language of our UI)
- etc.
I've directed a small handfull of admins to /register for creating admins.
We're about to release our Bulk Add/Create users tool.

What's REALLY missing now is UI that gets an admin from the dashboard to these 2 ways to create users. Currently, if an admin wants to create a user they still need to ask.

### PR DESCRIPTION
**Pull request by rmunn**
Fixes #666.

Extracted the relevant parts of the register page into a component and put it in an admin-only "Create User" modal. Help text at the top of the modal links to the relevant parts of the (new-ish) Scribe documentation.

## Help needed

Need to find the right CSS classes to format the text at the top of the modal so it looks good. Right now it's ugly: there's no blank space between the two paragraphs, for example.

## Screenshots

UI design on large screens:

image

UI design on small screens:

image

This is basically exactly the same as how the Projects tab and its "Create Project" button are presented.

Modal design:

image

### CODE REVIEW

**Comment by myieye** [comment]
@rmunn That button looks good 👍 

### But before you extract the register page!

There has been some discussion on this idea in our agenda doc.
I've [copied it to the issue](https://github.com/sillsdev/languageforge-lexbox/issues/666#issuecomment-2063417185) you're working on. Please take a look! It sounds like we don't actually want to extract/use the register page 😬

**Comment by rmunn** [comment]
After discussion with Kevin and Tim, implemented this as a way for admins to create a single **guest** user (will have the CreatedById field set to the admin's ID). That fits with the design I think Kevin was thinking of in https://github.com/sillsdev/languageforge-lexbox/issues/666#issuecomment-2083472063.

Other than a little help needed with the CSS classes for the instructions (see PR description), I think this is ready. Marking as ready for review.

File: backend/LexBoxApi/Controllers/UserController.cs [hunk 1]
```diff
@@ -80,20 +81,24 @@ public async Task<ActionResult<LexAuthUser>> RegisterAccount(RegisterAccountInpu
...
             IsAdmin = false,
             EmailVerified = emailVerified,
+            CreatedById = createdByAdmin ? jwtUser?.Id : null,
             Locked = false,
             CanCreateProjects = false
         };
         registerActivity?.AddTag("app.user.id", userEntity.Id);
         _lexBoxDbContext.Users.Add(userEntity);
-        if (jwtUser is not null && jwtUser.Projects.Length > 0)
+        if (jwtUser is not null && !createdByAdmin && jwtUser.Projects.Length > 0)
```
[the remark this answers is not in the document]
Comment (myieye) [reply backend/LexBoxApi/Controllers/UserController.cs:90]: I just changed this as I was already working on the PR. I'll let you resolve the comment.

**Review by myieye — APPROVED** [review]
Nice work 🙂 👍 
I pushed a commit where I played with the styles and wording.
I also fixed a bug that was already in the code.

Here's what it looks like now:
image
image

**Comment by hahn-kev** [comment]
Nice job! I hope this will help the admins out. 

I think it would make sense to remove turnstile from the dialog since we trust the user (an admin) in the dialog. This would require changing the controller code, but it would probably make more sense to make a separate controller action (or gql mutation) to create the user, that way we're not trying to fit all the logic into a single code path.

**Review by hahn-kev — CHANGES_REQUESTED** [review]
Nice job! I hope this will help the admins out.

I think it would make sense to remove turnstile from the dialog since we trust the user (an admin) in the dialog. This would require changing the controller code, but it would probably make more sense to make a separate controller action (or gql mutation) to create the user, that way we're not trying to fit all the logic into a single code path.

**Comment by myieye** [comment]
Yeah, we are playing with fire a bit having these 3 workflows in the same place (registration, invitations and guest user creation).

What would probably make most sense is something like this:
- Registration page can now only be accessed if you're not authenticated
- API endpoints:
   - Registration endpoint
     - REST API, so it's easy to mark it as [AllowAnonymous]. At least, I don't think we can mark specific GQL mutations with that.
     - Never looks for/uses a JWT token
   - Invitation endpoint
     - GQL mutation or REST API, doesn't really matter
     - Requires a Register audience token
   - Guest user creation endpoint
     - GQL mutation, so the new user lands in the GQL cache
     - Requires admin

**Comment by rmunn** [comment]
All right, I'll split the workflows in this PR. BTW, I've bumped the estimated size of the issue because it has grown so much during implementation.

**Comment by rmunn** [comment]
Changing this PR back to draft since it's now a WIP again.

File: backend/LexBoxApi/Controllers/UserController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs]: FYI, I added back the `jwtUser.Projects.Length > 0` check because in the future we might be issuing invitations that don't come attached to projects (e.g., inviting users to an organization).

File: frontend/src/lib/i18n/locales/en.json [hunk 2]
```diff
@@ -39,6 +39,11 @@
         }
       }
     },
+    "create_user_modal": {
+      "create_user": "Create User",
+      "help_create_single_guest_user": "Did you know you can also [create users on a project page]({addProjectMember})? That also allows you to make those users members of the project in one step; users created here will not be members of any project and will need a second step to add them to a project.",
+      "help_create_bulk_guest_users": "If you want to create multiple users at once, the best way is probably to use the [Bulk Add/Create Project Members]({bulkAddCreate}) tool.",
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/i18n/locales/en.json]: Huh. I just naturally assumed it was using the names. Thanks for fixing that; next time I'll explicitly pass in just the properties the translated text needs rather than passing in a big bag of properties.

File: backend/LexBoxApi/GraphQL/UserMutations.cs [hunk 3]
```diff
@@ -63,6 +74,50 @@ EmailService emailService
...
+    [AdminRequired]
+    public async Task<User> CreateGuestUserByAdmin(
+        LoggedInContext loggedInContext,
+        CreateGuestUserByAdminInput input,
+        LexBoxDbContext dbContext
+    )
+    {
+        using var createGuestUserActivity = LexBoxActivitySource.Get().StartActivity("CreateGuestUser");
+
+        var hasExistingUser = input.Email is null && input.Username is null
```
Comment (myieye) [inline backend/LexBoxApi/GraphQL/UserMutations.cs:90]: Supporting a username is going to mean extra work, but I agree that it makes sense, considering we're trying to extend/improve the experience of using the bulk tool.

I think there should probably just be one field in the UI for either the email or the username.
We're doing that in the Add-member modal and have validation that looks like this:
```
    usernameOrEmail: z.string().trim()
      .min(1, $t('project_page.add_user.empty_user_field'))
      .refine((value) => !value.includes('@') || isEmail(value), { message: $t('form.invalid_email') }),
```
(it's not complete enough for your case, because it doesn't have to validate usernames)

You could still split that field into 2 in the GQL input payload.

Anyway, I think there should only be 1 text field, however you end up doing this 🙂

**Comment by rmunn** [comment]
Changes are working, E2E tests are passing, no lint or svelte-check errors any more. Time for code review.

File: backend/LexBoxApi/Controllers/UserController.cs [hunk 4]
```diff
@@ -65,27 +65,54 @@ public async Task<ActionResult<LexAuthUser>> RegisterAccount(RegisterAccountInpu
...
+            ModelState.AddModelError<RegisterAccountInput>(r => r.TurnstileToken, "token invalid");
+            return ValidationProblem(ModelState);
+        }
+
+        var jwtUser = _loggedInContext.User;
+        if (jwtUser.Audience != LexboxAudience.RegisterAccount)
+        {
+            // TODO: Figure out how to register this error (AddModelError<RegisterAccountInput> isn't correct, obviously)
+            ModelState.AddModelError<RegisterAccountInput>(r => r.Email, "invitation page accessed via invalid JWT");
+        }
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs]: Already done before I saw this comment. :-)

File: backend/LexBoxApi/GraphQL/UserMutations.cs (hunk 3 above)
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/UserMutations.cs:90]: Admins can now create individual users with just a username in commit 6b19e6c2.

Comment (rmunn) [reply backend/LexBoxApi/GraphQL/UserMutations.cs:90]: Note that this doesn't validate the username, because we haven't decided on a rule for usernames. Should they be restricted to alphanumeric plus underscore only? Should uppercase be allowed? What about accents and/or non-Roman characters — what if someone wants a username in Thai or Korean?

Once we pick a rule, I'll implement that, but it's the end of the day and I'm not quite up to making that decision in my current state of mental fatigue. :-) Leaving this conversation unresolved for now, so that we can decide on what characters we'll allow in usernames.

**Comment by rmunn** [comment]
Okay, finished addressing all of Tim's review comments. Implemented UI for allowing usernames. I currently do **not** validate usernames because we haven't yet decided on a rule; I could say "only lowercase alphanumeric plus underscore", but that might be too restrictive. We might want to allow hyphens and/or dots, for example, and we might want to allow uppercase (though I'd rather not, so that we don't have to worry about case collisions between rmunn and Rmunn and RMunn). Anyway, this is ready to review now, even though the usernames aren't validating yet, because validating them is going to be a very minor change.

@hahn-kev - Want to give this a re-review?

**Comment by myieye** [comment]
Regarding login validation...
1) We actually do already have login validation in the bulk-create tool. We're using: `/^[a-zA-Z0-9_]+$/`.
2) The Username collumn uses a case-insensitive collation, so we can allow capitalization without getting duplicates
3) It [sounds like](https://www.redmine.org/issues/6166#:~:text=Redmine%20by%20default%20does%20neither,configured%20individually%20for%20each%20language.) this is what Redmine has: `/^[a-z0-9_\-@\.]*$/`. So funnily enough, they allow email addresses. Well, we're not going to do that 😆. It also means that users of LD-Redmine survived only being allowed to use latin characters. That's what our admins are used to.

So, I think it would be fine to just continue to use the pattern we alrady use for bulk-create.

File: frontend/src/lib/components/Users/CreateUser.svelte [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    const urlValues = getSearchParamValues<RegisterPageQueryParams>();
+    form.update((form) => {
+      if (urlValues.name) form.name = urlValues.name;
+      if (urlValues.email) form.email = urlValues.email;
+      return form;
+    }, { taint: true });
+  });
+</script>
+
+<ProtectedForm {enhance} bind:turnstileToken>
```
Comment (myieye) [inline frontend/src/lib/components/Users/CreateUser.svelte:72]: You're still using a turnstile protected form for creating guest users.
Maybe this would be a good use of `<svelte:component>` based on something like `export let protectedForm = false`;
(you can of course make `true` the default if you like. `false` as the default just makes things slightly "prettier" 😆)

File: frontend/src/lib/components/Users/CreateUserModal.svelte [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        />
+        <Markdown
+          md={$t('admin_dashboard.create_user_modal.help_create_bulk_guest_users', { helpLink: helpLinks.bulkAddCreate })}
+          plugins={[{ renderer: { a: NewTabLinkRenderer } }]}
+        />
+      </div>
+    </div>
+  </div>
+  <h1 class="text-center text-xl">{$t('admin_dashboard.create_user_modal.create_user')}</h1>
+  <CreateUser {endpoint} autoLogin={false} allowUsernames={true}
```
Comment (myieye) [inline frontend/src/lib/components/Users/CreateUserModal.svelte:36]: Prettiness 🤩 
```suggestion
  <CreateUser {endpoint} autoLogin={false} allowUsernames
```

**Review by myieye — CHANGES_REQUESTED** [review]
I really appreciate having 3 endpoints. Thank you.
It's weird....so much more code, but it's all simpler code.

File: backend/LexBoxApi/GraphQL/ProjectMutations.cs [hunk 7]
```diff
@@ -168,9 +174,14 @@ public record BulkAddProjectMembersResult(List<UserProjectRole> AddedMembers, Li
                 {
                     AddedMembers.Add(new UserProjectRole(user.Username ?? user.Email!, input.Role));
                     // Not yet a member, so add a membership. We don't want to touch existing memberships, which might have other roles
-                    user.Projects.Add(new ProjectUsers { Role = input.Role, ProjectId = input.ProjectId, UserId = user.Id });
+                    user.Projects.Add(new ProjectUsers { Role = input.Role, ProjectId = input.ProjectId.Value, UserId = user.Id });
                 }
             }
+            else
+            {
+                // No project ID specified, user already exists. This is probably part of bulk-adding through the admin dashboard or org page.
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/ProjectMutations.cs:182]: We should add a user filter for "user accounts I created". The URL query param could be `createdBy=email@example.com`. That would be useful not just in this scenario, but in several others, like removing guest user accounts after a workshop. Usually a workshop would involve just one project, so you could find them via the project, but it's entirely possible you might have multiple projects during a large workshop and then you might not remember all the usernames you created.

File: frontend/src/lib/components/Users/CreateUser.svelte (hunk 5 above)
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte:72]: Ah right, I was meaning to remove the turnstile from the CreateUserModal. Thanks for catching that.

File: frontend/src/lib/components/Users/CreateUserModal.svelte (hunk 6 above)
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUserModal.svelte:36]: Done in commit adebdc60 — I'd forgotten Svelte lets you do that with boolean props. Much nicer. Too bad there isn't a form like `!autoLogin`, but that would probably be too much parsing for too little gain for the Svelte team to consider that.

File: frontend/src/lib/components/Users/CreateUser.svelte (hunk 5 above)
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte]: Done in commit 0b9fd488.

File: frontend/src/lib/components/Users/CreateUser.svelte [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  import { register, acceptInvitation, createGuestUserByAdmin } from '$lib/user';
+  import { getSearchParamValues } from '$lib/util/query-params';
+  import { onMount } from 'svelte';
+  import { z } from 'zod';
+
+  export let autoLogin = true;
+  export let allowUsernames = false;
+  export let onSubmit: (() => void) | undefined = undefined;
+  export let submitButtonText = $t('register.button_register');
+  export let endpoint: 'register' | 'acceptInvitation' | 'createGuestUserByAdmin';
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte]: Done in commit 50fe0ce2.

File: backend/LexBoxApi/GraphQL/UserMutations.cs [hunk 9]
```diff
@@ -23,6 +27,13 @@ public record ChangeUserAccountBySelfInput(Guid UserId, string? Email, string Na
         : ChangeUserAccountDataInput(UserId, Email, Name);
     public record ChangeUserAccountByAdminInput(Guid UserId, string? Email, string Name, UserRole Role)
         : ChangeUserAccountDataInput(UserId, Email, Name);
+    public record CreateGuestUserByAdminInput(
+        string? Email,
+        string? Name,
+        string? Username,
+        string Locale,
+        string PasswordHash,
+        int? PasswordStrength);
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/UserMutations.cs]: Done in commit 57b45712.

File: backend/LexBoxApi/GraphQL/UserMutations.cs [hunk 10]
```diff
@@ -23,6 +27,13 @@ public record ChangeUserAccountBySelfInput(Guid UserId, string? Email, string Na
         : ChangeUserAccountDataInput(UserId, Email, Name);
     public record ChangeUserAccountByAdminInput(Guid UserId, string? Email, string Name, UserRole Role)
         : ChangeUserAccountDataInput(UserId, Email, Name);
+    public record CreateGuestUserByAdminInput(
+        string? Email,
+        string? Name,
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/UserMutations.cs]: Done in commit 328c3bcb.

File: backend/LexBoxApi/GraphQL/ProjectMutations.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/ProjectMutations.cs:182]: Apart from #810, not sure there's any required changes here at the moment. Leaving conversation open for now so Tim can more easily see it and reply if he wants to.

File: frontend/src/lib/components/Users/CreateUser.svelte [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      }
+      if (error.invalidInput) {
+        $errors.email = [$t('register.invalid_input')];
+        // If we later allow username-only accounts to be created, add error message to $errors.username as well and change message to "specify email or username"
+      }
+      return;
+    }
+    if (user) {
+      if (onSubmit) onSubmit();
+      if (autoLogin) await goto('/home', { invalidateAll: true }); // invalidate so we get the user from the server
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte]: Done. (Via multiple commits, not going to take time to list each commit hash here).

File: frontend/src/lib/components/Users/CreateUser.svelte [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    if (error) {
+      if (error.turnstile) {
+        $message = $t('turnstile.invalid');
+      }
+      if (error.accountExists) {
+        $errors.email = [$t('register.account_exists')];
+      }
+      if (error.invalidInput) {
+        $errors.email = [$t('register.invalid_input')];
+        // If we later allow username-only accounts to be created, add error message to $errors.username as well and change message to "specify email or username"
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte]: Done in commit 1e3cb445.

**Review by myieye — CHANGES_REQUESTED** [review]
I think there's just one missing piece.

File: frontend/src/lib/components/Users/CreateUser.svelte [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+  };
+  let turnstileToken = '';
+  // $locale is the locale that our i18n is using for them (i.e. the best available option we have for them)
+  // getLanguageCodeFromNavigator() gives us the language/locale they probably actually want. Maybe we'll support it in the future.
+  const userLocale = getLanguageCodeFromNavigator() ?? $locale;
+  const formSchema = z.object({
+    name: z.string().trim().min(1, $t('register.name_missing')),
+    email: z.string().trim()
+      .min(1, $t('project_page.add_user.empty_user_field'))
+      .refine((value) => isEmail(value) || (allowUsernames && usernameRe.test(value)), { message: $t('register.invalid_input') }),
```
Comment (myieye) [inline frontend/src/lib/components/Users/CreateUser.svelte:30]: I think it's important to tell the user what the username rules are. And the username validation warning should be seperate from the email warning. We already have a common email warning (`form.invalid_email`).

We already have a message for the bulk tool that includes the rules. For this I'd reword it to: `Invalid login/username. Only letters, numbers, and underscore (_) characters are allowed.`

Note: I've changed the wording of the second sentence. Please use either this wording or the current wording for both. Pick whichever you like most.

File: backend/LexBoxApi/GraphQL/ProjectMutations.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/LexBoxApi/GraphQL/ProjectMutations.cs:182]: I agree a org specific mutation would make sense once we get there, but I'm fine with leaving it if you don't want to remove it.

File: frontend/tests/pages/acceptInvitationPage.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+import { BasePage } from './basePage';
+
+export class AcceptInvitationPage extends BasePage {
+  constructor(page: Page) {
+    super(page, page.getByRole('heading', {name: 'Accept Invitation'}), `/acceptInvitation`);
+  }
+
+  async fillForm(name: string, email: string, password: string): Promise<void> {
+    await this.page.getByLabel('Name').fill(name);
+    await this.page.getByLabel('Email').fill(email);
```
Comment (hahn-kev) [inline frontend/tests/pages/acceptInvitationPage.ts:11]: this should be optional since the Accept invite page should have the email pre populated

**Review by hahn-kev — CHANGES_REQUESTED** [review]
nice job Robin! I was able to test it all out and didn't find any issues. I left some minor feedback.

Sorry it's taken me so long to get around to reviewing this PR for you.

File: frontend/src/lib/components/Users/CreateUser.svelte [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    const urlValues = getSearchParamValues<RegisterPageQueryParams>();
+    form.update((form) => {
+      if (urlValues.name) form.name = urlValues.name;
+      if (urlValues.email) form.email = urlValues.email;
+      return form;
+    }, { taint: true });
+  });
+</script>
+
+<svelte:component this={skipTurnstile ? Form : ProtectedForm} {enhance} bind:turnstileToken>
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte:66]: I was actually expecting a compile error because `turnstileToken` isn't a property on Form. We could, in theory, solve this by making a component called something like `MaybeProtectedForm` which looks like this:

```html
{#if skipTurnstile}
<Form {enhance}><slot /></Form>
{:else}
<ProtectedForm {enhance} bind:turnstileToken><slot /></ProtectedForm>
{/if}
```

But that gets a little ridiculous. The other alternative (besides leaving it as-is) would be to add `export let turnstileToken = ''` to the Form component, which I also don't like since it's not actually used in Form and would only be there to make this one line type-check (so to speak).

I'll open an issue to track this; I agree that I don't like letting warnings pile up ignored, because you soon start ignoring real problems, but I don't see a *good* solution right now.

File: frontend/src/lib/user.ts [hunk 16]
```diff
@@ -101,13 +105,56 @@ export async function register(password: string, passwordStrength: number, name:
...
+export async function createGuestUserByAdmin(password: string, passwordStrength: number, name: string, email: string, locale: string, _turnstileToken: string): Promise<RegisterResponse> {
+  const passwordHash = await hash(password);
+  const gqlInput: CreateGuestUserByAdminInput = {
+    passwordHash,
+    passwordStrength,
+    name,
+    locale,
+  };
+  if (email.includes('@')) {
+    gqlInput.email = email;
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/user.ts:130]: Same concerns would apply to the bulk-add tool once we decided to allow email addresses: we add the account and then send a verification email. This doesn't add anything new; if we need to make a change, we'll need to change how we handle the bulk-add tool as well. I'll open a separate issue to track that, because making a change to the bulk-add tool feels like scope creep for this specific PR.

File: frontend/src/lib/user.ts [hunk 17]
```diff
@@ -101,13 +105,56 @@ export async function register(password: string, passwordStrength: number, name:
...
+  const gqlResponse = await _createGuestUserByAdmin(gqlInput);
+  // const registerResponse = { error?: { turnstile: boolean, accountExists: boolean }, user?: LexAuthUser };
+  if (gqlResponse.error?.byType('UniqueValueError')) {
+    return { error: { invalidInput: false, turnstile: false, accountExists: true }};
+  }
+  if (gqlResponse.error?.byType('RequiredError')) {
+    return { error: { invalidInput: true, turnstile: false, accountExists: false }};
+  }
+  if (!gqlResponse.data?.createGuestUserByAdmin.lexAuthUser ) {
+    return { error: { invalidInput: true, turnstile: false, accountExists: false }};
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/user.ts:143]: Yeah, optional properties are probably the way to go here.

File: backend/LexBoxApi/GraphQL/UserMutations.cs [hunk 18]
```diff
@@ -63,6 +74,50 @@ EmailService emailService
...
+            PasswordStrength = UserService.ClampPasswordStrength(input.PasswordStrength),
+            IsAdmin = false,
+            EmailVerified = false,
+            CreatedById = admin.Id,
+            Locked = false,
+            CanCreateProjects = false
+        };
+        createGuestUserActivity?.AddTag("app.user.id", userEntity.Id);
+        dbContext.Users.Add(userEntity);
+        await dbContext.SaveChangesAsync();
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/UserMutations.cs:118]: Done in commit 024113c1.

File: frontend/src/lib/user.ts [hunk 19]
```diff
@@ -101,13 +105,56 @@ export async function register(password: string, passwordStrength: number, name:
...
+    name,
+    locale,
+  };
+  if (email.includes('@')) {
+    gqlInput.email = email;
+  } else {
+    gqlInput.username = email;
+  }
+  const gqlResponse = await _createGuestUserByAdmin(gqlInput);
+  // const registerResponse = { error?: { turnstile: boolean, accountExists: boolean }, user?: LexAuthUser };
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/user.ts]: Done in commit 024113c1.

File: frontend/src/lib/user.ts [hunk 20]
```diff
@@ -18,6 +19,7 @@ type RegisterResponseErrors = {
     /* eslint-disable @typescript-eslint/naming-convention */
     TurnstileToken?: unknown,
     Email?: unknown,
+    Required?: unknown,
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/user.ts]: Done in commit 024113c1.

File: frontend/src/routes/(authenticated)/admin/+page.svelte [hunk 21]
```diff
@@ -125,15 +127,13 @@
...
+            on:click={() => createUserModal.open()}>
             <span class="admin-tabs:hidden">
-              {$t('admin_dashboard.how_to_create_users')}
+              {$t('admin_dashboard.create_user_modal.create_user')}
             </span>
-            <span class="i-mdi-plus text-2xl group-hover:i-mdi-open-in-new" />
-          </a>
-        </div>
+            <span class="i-mdi-plus text-2xl" />
+          </button>        </div>
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/routes/(authenticated)/admin/+page.svelte]: Fixed in commit 024113c1.

File: frontend/tests/pages/acceptInvitationPage.ts (hunk 14 above)
Comment (rmunn) [reply frontend/tests/pages/acceptInvitationPage.ts]: Done in commit 024113c1.

File: frontend/src/lib/user.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/user.ts]: Done in commit e563e4a3.

File: frontend/src/lib/user.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/user.ts:130]: Created https://github.com/sillsdev/languageforge-lexbox/issues/815 to track this. Leaving this conversation open for now so Kevin and/or Tim can easily respond.

File: frontend/src/lib/components/Users/CreateUser.svelte (hunk 15 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte:66]: Actually, one good solution that just occurred to me would be to just continue using ProtectedForm even though we do trust our admins, because it doesn't cost them any effort. @myieye, you were the one who suggested removing the turnstile token for the admin-only create user modal. Would you be okay with reverting that change and leaving the turnstile token in the form even for admins? I don't think there's much drawback to doing that.

[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte:66]: Did a bit of digging into the Svelte 4 source and found that that warning is not suppressible with `<!-- svelte-ignore -->`, but that it only shows up in dev mode. So at this point, I'm thinking that perhaps that MaybeProtectedForm component is the way to go. It feels silly, but it's not all that large a component and it'll let us get rid of that error. Running out of time to do that today, though, so I'll have to do it tomorrow.

[the remark this answers is not in the document]
Comment (myieye) [reply frontend/src/lib/components/Users/CreateUser.svelte:66]: For the record, it was @hahn-kev that [brought up](https://github.com/sillsdev/languageforge-lexbox/pull/736#pullrequestreview-2055873940) removing turnstile, but I do agree.
I'd approve any of these options:
- Throwing away ProtectedForm and making turnstile an optional feature of Form
- Introducing a new component to pick between the two form components
- Living with the warning

[the remark this answers is not in the document]
Comment (hahn-kev) [reply frontend/src/lib/components/Users/CreateUser.svelte:66]: I like the first 2 options. But let's move this into it's own issue since it's not a simple solve on this already large PR

**Review by hahn-kev — APPROVED** [review]
Looks good to me, though it looks like we're still pending @myieye

File: frontend/src/lib/components/Users/CreateUser.svelte (hunk 13 above)
Comment (rmunn) [reply frontend/src/lib/components/Users/CreateUser.svelte]: Done in commit f027c6bc. Is that what you had in mind?

**Comment by rmunn** [comment]
@myieye - Commit f027c6bc should address your review comments. I left that conversation open in case what I did wasn't what you had in mind. Let me know. Also, I merged develop into this branch to fix the merge conflict in en.json resulting from that last change.

Once you're happy with the error message, I'm looking forward to finally getting this merged.

File: frontend/src/lib/components/Users/CreateUser.svelte (hunk 13 above)
Comment (myieye) [reply frontend/src/lib/components/Users/CreateUser.svelte]: It was a bit buggy, so I added [f2d011a](https://github.com/sillsdev/languageforge-lexbox/pull/736/commits/f2d011a30b27f1d918f3b7bbbd3d994727c6cde6).

**Review by myieye — APPROVED** [review]
Well that was an adventure 😬.
Good work hanging in there @rmunn! 👏 🚀
